internal/token/token.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2023 Chainguard, Inc.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package token
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"chainguard.dev/sdk/auth/login"
+	sdktoken "chainguard.dev/sdk/auth/token"
+	"chainguard.dev/sdk/sts"
+)
+
+const (
+	// auth0ClientID is the oauth2 clientID to indicate to the issuer to
+	// authenticate with Auth0.
+	auth0ClientID = "auth0"
+)
+
+var lock sync.Mutex
+
+// Get retrieves a Chainguard token, refreshing it if expired/non-existent or forceRefresh == true.
+// If automatic authentication is disabled, returns an unauthenticated error.
+func Get(ctx context.Context, cfg LoginConfig, forceRefresh bool) ([]byte, error) {
+	// Lock access to the token so no other process attempts to read or refresh
+	// while refreshing the token.
+	lock.Lock()
+	defer lock.Unlock()
```

Comment:
Since we use the mutex for read & write, it'd be better to use https://pkg.go.dev/sync#RWMutex

This will allow reads in parallel and will acquire a unique lock for write!